physical/oci_objectstorage/oci_objectstorage_ha.go

```diff
@@ -0,0 +1,554 @@
+// Copyright © 2019, Oracle and/or its affiliates.
+package oci_objectStorage
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/sdk/physical"
+	"github.com/oracle/oci-go-sdk/objectstorage"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// The lock implementation below prioritizes ensuring that there are no 2 primary at any given point in time
+// over high availability of the primary instance
+
+// Verify Backend satisfies the correct interfaces
+var _ physical.HABackend = (*Backend)(nil)
+var _ physical.Lock = (*Lock)(nil)
+
+const (
+	// LockRenewInterval is the time to wait between lock renewals.
+	LockRenewInterval = 3 * time.Second
+
+	// LockRetryInterval is the amount of time to wait if the lock fails before trying again.
+	LockRetryInterval = 5 * time.Second
+
+	// LockWatchRetryInterval is the amount of time to wait if a watch fails before trying again.
+	LockWatchRetryInterval = 2 * time.Second
+
+	// LockTTL is the default lock TTL.
+	LockTTL = 15 * time.Second
+
+	// LockWatchRetryMax is the number of times to retry a failed watch before signaling that leadership is lost.
+	LockWatchRetryMax = 4
+
+	// LockCacheMinAcceptableAge is minimum cache age in seconds to determine that its safe for a secondary instance
+	// to acquire lock.
+	LockCacheMinAcceptableAge = 45 * time.Second
+
+	// LockWriteRetriesOnFailures is the number of retries that are made on write 5xx failures.
+	LockWriteRetriesOnFailures = 4
+
+	ObjectStorageCallsReadTimeout = 3 * time.Second
+
+	ObjectStorageCallsWriteTimeout = 3 * time.Second
+)
+
+type LockCache struct {
+	// ETag values are unique identifiers generated by the OCI service and changed every time the object is modified.
+	etag       string
+	time       time.Time
+	lockRecord *LockRecord
+}
+
+type Lock struct {
+	// backend is the underlying physical backend.
+	backend *Backend
+
+	// Key is the name of the Key. Value is the Value of the Key.
+	key, value string
+
+	// held is a boolean indicating if the lock is currently held.
+	held bool
+
+	// Identity is the internal Identity of this Key (unique to this server instance).
+	identity string
+
+	internalLock sync.Mutex
+
+	// stopCh is the channel that stops all operations. It may be closed in the
+	// event of a leader loss or graceful shutdown. stopped is a boolean
+	// indicating if we are stopped - it exists to prevent double closing the
+	// channel. stopLock is a mutex around the locks.
+	stopCh   chan struct{}
+	stopped  bool
+	stopLock sync.Mutex
+
+	lockRecordCache atomic.Value
```

Comment:
Can you elaborate on why this cache value (and the logic around it) is necessary?

Reply:
@kalafut Ankush in our team who is author of this plugin is writing up details for this decision of using cache soon today.

Reply:
From Ankush (His github account is still pending to approval):
**Overview**
This document describes the HA implementation of hashicorp vault plugin using OCI objectstorage as the backend. The algorithm prioritizes ensuring that there are not 2 primary nodes at any given point over high availability of the primary instance. This algorithm may not work with byzantine clients.  

**Algorithm**
An implementation of lock is simulated using the objectstorage backend which is described in the following sections. Lock can only be acquired by one node at any given point in time. The node owning the lock assumes it-selves as the primary node.  All peer nodes not owning the lock assumes themselves as secondary nodes.  Peer nodes periodically monitor the lock and compete to acquire it if the current primary node fails to renew the lock within a specified period of time say x. The current primary node steps down to the secondary role if it is unable to renew the lock within a specified period of time say y.  The value of y has to be less than x. 

The lock is implemented as an object in object storage. The value of the object contains the identity of the owner of the lock. The ownership of the lock is assumed by the node, if it is able to successfully write to the object. Object storage maintains an Etag which is updated on every write operation. Object storage provides the ability to perform atomic compare and swap operations on an object level using Etags. This fulfills the basic requirement from the storage backend to ensure only one node acquires the lock successfully at any given point in time.  

All initially booted vault instances assume secondary status. If the lock object is not present, all nodes compete to acquire the lock and only one of them successfully acquires it and assumes it-selves as the primary node. The primary node re-writes the lock object every x seconds.  In reality, nodes can have skewed clocks and so absolute time from any of the nodes cannot be used to determine the last updated time of the lock by its owner. Instead, all nodes cache the value of the lock (identity of the current primary and the etag value) along with its local timestamp at which the local cache was updated or created. The peer nodes acting as secondary nodes, periodically monitor the lock for changes. If the lock gets updated, nodes update their cache along with its local timestamp at which this cached value was refreshed. If a node observes no changes to the lock from its cached value for a period z, where z > x, it competes to acquire the lock. On successful acquisition of lock, the node assumes it-selves as primary. Since z > x, it ensures that there are not two primaries at any given point in time. The times, x, y and z should be set far apart to factor the possibility of clock drift introduced by NTP.

**Notes**
Primary will lose the lock (after LockTTL=15 seconds on failure to re-write the lock) OR (failure to watch the lock after (LockWatchRetryMax=3 * LockWatchRetryInterval=2) seconds) whichever is earlier. Secondary instance acquires lock only after (LockCacheMinAcceptableAge=45) seconds. This provides margin for clock drift to prevent multiple primary scenario.
The maximum time for secondary to acquire a lock after primary dies is:
LockRetryInterval (5 seconds)+
LockCacheMinAcceptableAge (45 seconds) +
LockRetryInterval  (5 seconds) (assuming the cache matured just after when last check time secondary instance checked the lock) +
ProcessingTime (put/get + other processing)

Reply:
Thanks for this explanation. I've thought about it some and understand what you're going for here. I will caution that the HA logic is always tricky and has been a source of subtle bugs, which is why I commented about this approach, which brings in more complexity. One thing you may want to consider is some additional OCI-custom HA test cases. You can look at Dynamo for an example of what I mean. The generic HA tests know nothing about the implementation and can't really stress low level edge cases. But you could, for example, have a custom test that does things like immitates clock skew by altering the stored object's time, just to see whether the design assumptions hold.

This is not a blocker, but it is something I'd recommend looking into as a supplemental PR.